src/data/matrix/rank.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import linear_algebra.free_module.finite.rank
+
+/-!
+# Rank of matrices
+
+The rank of a matrix `A` is defined to be the rank of range of the linear map corresponding to `A`.
+This definition does not depend on the choice of basis, see `matrix.rank_eq_finrank_range_to_lin`.
+
+## Main declarations
+
+* `matrix.rank`: the rank of a matrix
+
+## TODO
+
+* Show that `matrix.rank` is equal to the row-rank and column-rank
+* Generalize away from fields
+
+-/
+
+open_locale matrix
+
+namespace matrix
+
+open finite_dimensional
+
+variables {m n o K : Type*} [m_fin : fintype m] [fintype n] [fintype o]
+variables [decidable_eq n] [decidable_eq o] [field K]
+variables (A : matrix m n K)
+
+/-- The rank of a matrix is the rank of its image. -/
+noncomputable def rank : ℕ := finrank K A.to_lin'.range
+
+@[simp] lemma rank_one : rank (1 : matrix n n K) = fintype.card n :=
+by rw [rank, to_lin'_one, linear_map.range_id, finrank_top, module.free.finrank_pi]
+
+@[simp] lemma rank_zero : rank (0 : matrix n n K) = 0 :=
+by rw [rank, linear_equiv.map_zero, linear_map.range_zero, finrank_bot]
+
+lemma rank_le_card_width : A.rank ≤ fintype.card n :=
+begin
+  convert nat.le_of_add_le_left (A.to_lin'.finrank_range_add_finrank_ker).le,
+  exact (module.free.finrank_pi K).symm,
+end
+
+lemma rank_le_width {m n : ℕ} (A : matrix (fin m) (fin n) K) : A.rank ≤ n :=
+A.rank_le_card_width.trans $ (fintype.card_fin n).le
+
+lemma rank_mul_le (B : matrix n o K) : (A ⬝ B).rank ≤ A.rank :=
+begin
+  refine linear_map.finrank_le_finrank_of_injective (submodule.of_le_injective _),
+  rw [to_lin'_mul],
+  exact linear_map.range_comp_le_range _ _,
+end
+
+lemma rank_of_invertible (A : matrix n n K) [invertible A] :
```

Comment:
I think we usually only use an `invertible` assumption if the lemma statement actually needs it:
```suggestion
lemma rank_of_invertible (A : matrix n n K) (h : is_unit A) :
```